### DISCUSSION TITLE
AP2 build also works on Raspberry Pi Zero W

### DISCUSSION THREAD

**Discussion by Siegum** [mikebrady/shairport-sync#1229] [General]
Hello all, I would also like to applaud all the developers involved in the new AP2 build. 

In one of the comments it is stated that "The AirPlay 2 build requires a good deal of extra library support and may not fit into smaller devices. It also requires more CPU power and more RAM.". Furthermore, a RPi3 is suggested as the minimum requirement.

This makes a lot of sense, but as I had a spare RPi Zero W, I decided to try out the AP2 build on this unit.

I did a clean install of Raspbian Buster Lite on a small (4GB) SD-card. 

The Zero is equipped with a DAC HAT (Phat Dac from Pimoroni), and the [driver for the HAT was installed](https://learn.pimoroni.com/tutorial/phat/raspberry-pi-phat-dac-install).
![image](https://user-images.githubusercontent.com/87268855/125197759-3db56280-e25f-11eb-9e83-fae3d3699d46.jpeg) ![image](https://user-images.githubusercontent.com/87268855/125197775-4c9c1500-e25f-11eb-9c1c-a62d958503c9.jpeg)

I then followed the instructions in [Building Shairport Sync for AirPlay 2](https://github.com/mikebrady/shairport-sync/blob/development/BUILDFORAP2.md). Everything installed without issue.

I have previously used the excellent [rpi-audio-receiver](https://github.com/nicokaiser/rpi-audio-receiver), and based on experience from this software, I [disabled on-board bluetooth](https://di-marco.net/blog/it/2020-04-18-tips-disabling_bluetooth_on_raspberry_pi/). I have previously experienced that IPv6 caused problems with Airplay 2 using Apple Airport Express A1392 units, so I also [disabled IPv6](https://www.howtoraspberry.com/2020/04/disable-ipv6-on-raspberry-pi/).

I then tried to airplay to two units simultaneously, one airport express (Main Stereo) and the Pi Zero (Pidac), and I was able to play without issue and in sync.
![image](https://user-images.githubusercontent.com/87268855/125197927-fbd8ec00-e25f-11eb-982f-c99b1f4454d3.jpeg)

I was curious if the Pi Zero was somehow pushed to the limit, so I installed [RPI-Monitor](https://github.com/XavierBerger/RPi-Monitor), to monitor the load while playing to the Pidac. Just after booting the Pi, and starting to play, this was the situation.
![image](https://user-images.githubusercontent.com/87268855/125197969-275bd680-e260-11eb-8813-7ff4de7cce57.png)

After playing for an hour, the load looked like this.
![image](https://user-images.githubusercontent.com/87268855/125198001-5bcf9280-e260-11eb-8b0a-58993d083601.png)
The load seems to have stabilized, and neither the load or the CPU temperature appears extreme, so my initial conclusion is that this will work perfectly on a RPi Zero W……very good job indeed.

**Comment by Siegum** [comment]
The screenshots of the RPI-monitor were a little puny, these should be better
![image](https://user-images.githubusercontent.com/87268855/125198577-aeaa4980-e262-11eb-843d-4897656eadd8.png)


![image](https://user-images.githubusercontent.com/87268855/125198596-c7b2fa80-e262-11eb-8477-431929bab374.png)

**Reply by charlesomer** [reply]
Thank you for testing this! As you've discovered, it is somewhat possible to run Shairport Sync with AirPlay 2 on a Pi Zero but we've noticed it does have its limitations. For example, scrubbing the volume quickly can cause dropouts in audio and sometimes skipping tracks can result in a quick dropout (at least for me it does). But, it is surprising just how well it can work and it's good you've got it working well for you! 😄

**Reply by Siegum** [reply]
Good to know, I will try and reproduce that error. Was is from IOS or MacOS that you have seen this ?

**Reply by charlesomer** [reply]
iOS, I haven't tried macOS with my Pi Zero yet so not sure if it experiences a similar thing

**Reply by ckdo** [reply]
Good to know ! I'll give it a try on my SolidRun Cubox (original) !

**Reply by gestcodev** [reply]
Hice todo el proceso y funciona bien, pero me gustaria saber si se puede controlar el audio desde otro dispositivo, algo asi como que sea independiente. Es lo que hace hoy el homepod mini.

Sorry mi english, i am LA

**Comment by mikebrady** [comment]
Many thanks for testing, @Siegum. I think the dropout @charlesomer is referring to _may_ be caused by the Pi's own built-in DAC driver, so if you're not using that, you might be okay.

Earlier versions of the AP2 support certainly did overwhelm the Pi Zero, so we want to be conservative in our guidance until we know better.

**Comment by noelhibbard** [comment]
I was surprised at how little load was on my RPi3B. I'm not surprised that the PiZero can also handle it. I'm using a USB DAC, maybe using the onboard DAC would add more load? The onboard DAC sounds terrible so I tried to avoid it.

Thanks for the heads up on that DAC. $14 is a good deal. My only issue with the PiZero is the added WiFi chatter.

**Comment by iChUdY** [comment]
Last dev works perfect on pi zero 👌🏻👌🏻

**Comment by colourofsound** [comment]
I have a PiZeroWH which I've just setup for Airplay 2. It has Pirate Audio's DAC Shim installed. I disable onboard audio when I have a DAC in the mix.

It works pretty well until you do any of the following:

- Pause
- De-select it from the Airplay menu on your streaming device and then reselecting

On doing either of these things, audio ceases to go to Shairport-Sync and this is reflected in the CPU usage; when streaming its around 30% whilst Shairport-Sync idles at about 7% when not receiving the audio.

You can re-select the device but pressing play yields no audio. Sometimes the source won't allow the music to be played at all.

The fix appears to be restarting the Service, but this only works about 75% of the time. I'm going to do the tweaks listed above and see if that makes a difference.

I should note: I had normal Shairport-Sync on this unit and removed it to put the Airplay2 version on. In order to be absolutely sure it wasnt a configuration issue from the lingering installation, and I rebuilt the PiZero from scratch and install the Airplay2 version fresh; and the issue persists.

**Reply by mikebrady** [reply]
Thanks for the report. Can you say what the sources are, please? Are they iOS or macOS or what? And would any of them be beta versions?

**Reply by colourofsound** [reply]
Okay, so just ran some tests:

From iPhone 12 Mini on latest iOS 15; Apple Music:
- Play and Pause works fine.
- Disconnecting from the receiver and then reconnecting results in no audio, but the music plays.

From Macbook Pro 2015 on latest Big Sur; Apple Music:
- Connecting to the receiver results in the music refusing to play at all on the player; you can hit play and scrub but it will not actually play or move forward through the track

Interestingly I also have a Pi 2 Model B with a HiFiBerry Amp Hat on, and it does not display these issues. The only odd thing that does is put the volume to zero the moment you touch the slider on the lock screen widget; using the volume buttons on the side of the phone alleviates this.

EDIT: It looks like the audio mixer is the issue:

```
pi@PiOffice:~ $ sudo service shairport-sync status
● shairport-sync.service - Shairport Sync - AirPlay Audio Receiver
   Loaded: loaded (/lib/systemd/system/shairport-sync.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2021-10-13 21:51:57 BST; 2min 6s ago
 Main PID: 391 (shairport-sync)
    Tasks: 19 (limit: 877)
   CGroup: /system.slice/shairport-sync.service
           └─391 /usr/local/bin/shairport-sync

Oct 13 21:51:57 PiOffice systemd[1]: Started Shairport Sync - AirPlay Audio Receiver.
Oct 13 21:52:54 PiOffice shairport-sync[391]: warning: failed to find mixer control "snd_rpi_hifiberry_dac".
```

So shairport-sync is intermittently kicking the mixer for some reason? Will need to do more tests I think

**Reply by colourofsound** [reply]
Commenting out the mixer control in shairport-sync.conf fixed the issue on iOS temporarily; but its now returned.

Still won't play from macOS at all.

**Reply by mikebrady** [reply]
I'm using a Pi 4 with a Pirate Audio's DAC Shim without issue, but only after I fixed the shim in place with standoff bolts. Here are a few pointers:
* The onboard audio DAC is not disabled -- there is no need.
* The overlay used is `hifiberry-dac`.

Thus, the relevant part of the `/boot/config.txt` file is:
```
# Additional overlays and parameters are documented /boot/overlays/README

# Enable audio (loads snd_bcm2835)
dtparam=audio=on
dtoverlay=hifiberry-dac
```
* The Pirate Audio's DAC Shim has no mixer, so it's an error to specify one. (You can use `alsamixer` to verify this.)

You can get a list of the real alsa hardware audio devices at the end of `shairport-sync -h`:
```
...
Available audio backends:
    alsa (default)

Settings and options for the audio backend "alsa":
    -d output-device    set the output device, default is "default".
    -c mixer-control    set the mixer control name, default is to use no mixer.
    -m mixer-device     set the mixer device, default is the output device.
    -i mixer-index      set the mixer index, default is 0.
    hardware output devices:
      "hw:Headphones"
      "hw:sndrpihifiberry"
```
Therefore, the setting for the output device in the configuration file should look like this:
```
...
// These are parameters for the "alsa" audio back end.
// For this section to be operative, Shairport Sync must be built with the following configuration flag:
// --with-alsa
alsa =
{
        output_device = "hw:sndrpihifiberry";
… 
```

That's what works for me.

**Reply by colourofsound** [reply]
Yeah mine is basically set up exactly like that, save for the disabled onboard audio which I have re-enabled for completionist sake.

Does Shairport have any verbose logging? Forgive if its documented; I can't find it if its detailed

**Reply by colourofsound** [reply]
Also - it will not play from macOS at all. The RPi 2B also running the Airplay2 version of Shairport has no issues with macOS or iOS.

Theres something inherent in the Pi Zero causing the issue, it seems to me

**Reply by colourofsound** [reply]
Haha - the plot thickens! I can stream to the PiZero if I set the _System Audio_ to the PiZero - but it will not stream music if selected directly inside the Apple Music app.

**Reply by colourofsound** [reply]
Okay, so to summarise:
- I can Airplay 2 to the PiZero over iOS, once. On the second time of connecting the music is playing but there is no audio heard.
- I can Airplay 2 to the PiZero over macOS, but only if the PiZero is selected as the main audio output in System Preferences. This works perfectly - I can select and de-select it, pause, and it all plays fine.
- Streaming from macOS from within the Apple Music app does not work - the music will not even start once the PiZero is selected.
- The RPi 2B which has a HiFiBerryAmp connected has none of these issues whatsoever - will happily stream from iOS, Apple Music on macOS and System Audio on macOS.

Sorry for the repeat replies!

**Reply by noelhibbard** [reply]
It would be interesting to see the config from both RPi's along with the output from ```shairport-sync --version``` on both.

**Reply by colourofsound** [reply]
I'd altered the RPi2B config...but hadn't actually uncommented the output device and mixer that I'd edited 🤦 so in actual fact that was running a fully default config. I'd disabled the onboard audio so I guess this makes sense why it just worked - no other audio device was available.

The config on the Zero is default, save for this line: `output_device = "hw:sndrpihifiberry";` being altered and uncommented.
However, I've now commented out the device, and disabled the onboard sound...and its now working a little more reliably; however there is still some stuttering when starting the stream and it doesnt always play. Better than before though.

Both RPis are on version `4.1-dev-8-gd25a20aa-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc`

macOS remains the same - fine as System Audio; won't stream direct from Apple Music

**Reply by mikebrady** [reply]
> Also - it will not play from macOS at all. The RPi 2B also running the Airplay2 version of Shairport has no issues with macOS or iOS.
> 
> Theres something inherent in the Pi Zero causing the issue, it seems to me

The RPi 2B is more powerful than the Pi Zero -- in fact it's about four times more powerful. Recall than for AP2, Shairport Sync recommends an RPi 3 or better. 

That said, it does work with a Pi Zero, as other contributors have noted.

Are you running Raspberry Pi OS or Raspberry Pi OS Lite.

> Does Shairport have any verbose logging? Forgive if its documented; I can't find it if its detailed

Yes, it's detailed on the man page and in the configuration file itself.

**Reply by colourofsound** [reply]
> That said, it does work with a Pi Zero, as other contributors have noted.

It’s only taking about 40% CPU when operating okay; which as it’s only function is perfectly acceptable.

I assume the things I’m experiencing are a mixture of the experimental nature of it plus the limitations of the Zero. 

> Are you running Raspberry Pi OS or Raspberry Pi OS Lite.

Lite on all my Pis

It is odd though how it won’t stream from Apple Music at all - that seems like a software issue rather than a hardware limitation. 

All good fun though 😀

**Comment by Siegum** [comment]
I am not the developer of shairport-sync, just the threadstarter, but here are my two cents on the issues that Chris Walker have reported.
**1. Shairport-sync does not play music on IOS when you disconnect and reconnect.**
 I can partly reproduce this problem, but if I disconnect and reconnect repeatedly (4-5 times), then suddenly it starts playing again. If I do not connect/disconnect repeatedly, I observe the same as Chris....namely that no sound comes from the pi zero
**2. Shairport-sync will not play from the Apple Music, unless you select the pi as System Audio on macOS.**
Again I can reproduce this problem.....but if I first select another source (in my case an ATV4K) and then select the pi as a second source it will start playing. After this I can deselect the atv4k and the music plays only on the pi from within the Apple Music interface. At this point I can select and deselect the pi at will with no issues.

I agree that these bugs could arise from either i)piZeroW, ii)Shairport-sync or iii)IOS/macOS, but it seems both bugs can be "fixed" without restarting the service.

**Reply by Nosleep28** [reply]
I do confirm **1. Shairport-sync does not play music on IOS when you disconnect and reconnect.** having the same three PI Zero WH with HifiBerry and Lite install.

**Reply by Macelovic** [reply]
I think Im pretty late to the game but Ive got AP2 running on my Pi Zero without any issues on DietPi yesterday. Today Im coming back from work and wanted to play some music and nothing...  reconnecting repeatedly did not work. Kinda struggling getting it back to work. Think I might try a Pi Zero 2

**Reply by mikebrady** [reply]
Thanks. Let me suggest that you make sure to use the latest version of Shairport Sync. I've been using it on a Pi Zero W for a little while and it my experience is that it works pretty well almost all the time.

**Comment by mikebrady** [comment]
Thanks for all the interesting observations.

There's a bit more to add to it. At the moment, there are two types of AirPlay 2 streams — Realtime and Buffered. The Realtime stream is ALAC-encoded and the Buffered stream is AAC-encoded. The development version of Shairport Sync supports both.

Different apps and modes select a stream type automatically. Apple Music on iOS and on macOS uses Buffered Audio, thus AAC. System Audio on macOS, Tunein Radio and some others use Realtime Audio, thus ALAC.

Decoding ALAC seems to take considerably less CPU effort than decoding AAC. In fact, Realtime Audio streams will play on an original Pi where Buffered Audio streams will not work at all. It's also worth remembering that the 40% CPU load when working normally on a Pi Zero is an average figure. 

So, my feeling is that the Pi Zero is able to keep up with a Realtime Audio stream rather easily but, for some reason, is only able to keep up with a Buffered Audio stream under favourable conditions. There may be some peculiarity of the Buffered Audio processing chain that makes it particularly susceptible to delays and glitches, but I haven't been able to find it.

Two things that might be worth experimenting with if you wish are:
1. Set the interpolation mode to `basic` to avoid the possibility that `soxr`-based interpolation, chosen automatically, is bogging the system down.
2. Set the compilation flags to do the highest level of optimisation and see if it makes a difference. Something like:
```
CFLAGS="-O3 -g" CXXFLAGS="-O3 -g" ./configure ...
```
Not sure it makes much difference, TBH, but worth trying.

Definitely good fun alright 🙂.

**Reply by Nosleep28** [reply]
I just checked your recommendations about optimizing, and I have good results with the following:  

1. I switch to `apple` as an `alac_decoder` lib  and compiled your ALAC with the option `CFLAGS="-O3 -g" CXXFLAGS="-O3 -g" ./configure` ... without really understanding what these options are doing, just hoping that these have a general meaning for the compiler 

2. `interpolation_mode` to `basic` as recommended 3. Compiled shairport-sync with `CFLAGS="-O3 -g" CXXFLAGS="-O3 -g" ./configure --sysconfdir=/etc --with-alsa     --with-soxr --with-avahi --with-ssl=openssl --with-systemd --with-airplay-2 --with-apple-alac`  

The result is definitely much better. Now the "stop" and "continue", "skip song", and volume control worked like a charm. Stressing the active array with changes worked perfectly and lead to a smile on my face  ;)   

But I have weired problems with syncing the third device. The first time after reboot, all my three devices syncing together. After a while (or in the next session) randomly one of the devices failed to play music synced. On iOS the device seems to be integrated in the array and playing normally. A few seconds later and a few tries of reconnecting, the device is no longer available. The systemctrl is reporting an active and healthy service. The logs do not say anything. Is there an option to extend the log information to trace level?  

Additionally, I noticed session interruption by another iOS client was possible, although `allow_session_interruption`  was on default (`"no"`) but this is a minor concern for me.

->please note that I have an hifiberry as an dac and I'n not using onboard dac. This might help the zero wh to handle the workload.

**Comment by Siegum** [comment]
The new Zero 2 W works without ANY of the above mentioned problems. I think that corroborates Mike Bradys suggestion that the original Pi Zero W is struggling with some peculiarities from the buffered audio stream. So if you want AP2 in the small form factor, I think it is a no-brainer to suggest the new Pi Zero 2 W  :-)

**Reply by colourofsound** [reply]
Very interesting! Sadly I have reverted back to the master branch on both my Zero and my RPi 2B as the newest iOS version seems to break streaming to them. I couldn't parse from the logs what the issue was; and on balance reliability is higher up my list than multi-room.

**Reply by mikebrady** [reply]
Thanks for the observations. FWIW, there was a recent update to the development branch to keep it working with iOS 15.1. I have a PiZ2W on order 😊!

**Reply by joe-a-b** [reply]
I'm interested if you have as success with your Pi Zero 2 board as I did!   I received mine this month, recently tried replacing my older pi zero (that was unable to keep up cpu with both lighshowpi running the shairport-sync, it would lag horribly).  Happy to report that the surgery was a success and sound is no longer _as_ garbled _as it was_ .   shairport runs from inside lightshowpi --with-stdout (using a usb hat https://makerspot.com/docking-hub-for-raspberry-pi-zero/).   running shairport first, sound is clean.   i'm still noticing garbling but not sure cause , it happens even if no output on pins but with lightshowpi running to virtual pins.

**Reply by mikebrady** [reply]
>   i'm still noticing garbling but not sure cause , it happens even if no output on pins but with lightshowpi running to virtual pins. 

That's interesting alright. Can I ask what program is responsible for finally sending the audio to output? Is it the lightshowpi program itself?

**Reply by joe-a-b** [reply]
Thank you!  I think its through the shairport_sync .   not critical to me, it will be outdoors, i tried to get sound of the [pitch difference, apologize for poor demo](https://youtu.be/5m3C975MohY).
It's Alsa output as fars I can tell through the usb card 1.
```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: vc4hdmi [vc4-hdmi], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
**card 1**: Device [USB Audio Device], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
`/usr/share/alsa/alas.conf`:

```
defaults.ctl.card 1
defaults.pcm.card 1

```

```
$ shairport-sync -h
…
…
Available audio backends:
    alsa (default)
    pa
    pipe
    stdout

Settings and options for the audio backend "alsa":
    -d output-device    set the output device, default is "default".
    -c mixer-control    set the mixer control name, default is to use no mixer.
    -m mixer-device     set the mixer device, default is the output device.
    -i mixer-index      set the mixer index, default is 0.
    hardware output devices:
      "hw:Device"
      "hw:vc4hdmi"
```

```
$ lsusb
Bus 001 Device 004: ID 0bda:8152 Realtek Semiconductor Corp. RTL8152 Fast Ethernet Adapter
Bus 001 Device 003: ID 0d8c:0014 C-Media Electronics, Inc. Audio Adapter (Unitek Y-247A)
Bus 001 Device 002: ID 1a40:0201 Terminus Technology Inc. FE 2.1 7-port Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

**Reply by joe-a-b** [reply]
> > i'm still noticing garbling but not sure cause , it happens even if no output on pins but with lightshowpi running to virtual pins.
> 
> That's interesting alright. Can I ask what program is responsible for finally sending the audio to output? Is it the lightshowpi program itself?

Ok feeling dumber on this day, than others - Misconfiguration on my part!  I rechecked lightshowpi defaults.cfg and it had sample rate different than the other boxes. (48k, vs 44.1k on the raspi4b) i think that explains the pitch difference!  

Thanks again for listening.  Shairport sync is an awesome tool!

**Comment by quantonian** [comment]
I wanted to share my experience with AP2 and the pi zero 2 w as well as extend my appreciation to Mike and everyone who has contributed to getting AP2 up and running!

I was able to get AP2 running perfectly on the pi zero 2 using a very inexpensive PCM5102A DAC ($10-$12 on ebay or amazon). I'm using the latest raspberry pi OS. 

I have several hifiberry DACs (DAC+ Standard & Zero) and the sound with this inexpensive DAC is identical (Hifiberry uses the PCM5122 chip - the only difference is the the 5122 has hardware volume control which I don't need).

![image_123927839](https://user-images.githubusercontent.com/94640977/144476246-ce3db967-840d-4616-81ef-4dc61044d281.JPG)
![image](https://user-images.githubusercontent.com/94640977/144480031-93fb1d2e-b446-4df2-bc68-eb5c79408de3.png)

gpio configuration for DAC:
```
SCK        	 > Not wired 	(Internally generated) 
BCK       	 > PIN 12    	(GPIO18)
DIN        	 > PIN 40    	(GPIO21)
LRCK      	 > PIN 35    	(GPIO19)
GND       	 > PIN 6    	(GND) Ground
VIN       	 > PIN 2     	(5V)
```
Pi configuration:

Add the following to `/boot/config.txt`:
```
dtoverlay=hifiberry-dac
```

Here is the contents of  `/etc/shairport-sync.conf`:
```
// Sample Configuration File for Shairport Sync on a Raspberry Pi using the built-in audio DAC
general =
{
  volume_range_db = 60;
};

alsa =
{
  output_device = "hw:sndrpihifiberry";
};
```
Reboot and you are good to go!

Interestingly, the `aplay -L` shows this DAC as a hifiberry DAC:
```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_dac], device 0: HifiBerry DAC HiFi pcm5102a-hifi-0 [HifiBerry DAC HiFi pcm5102a-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
`top` shows about 22% cpu. 
![image](https://user-images.githubusercontent.com/94640977/144476402-d1b037d1-3767-4458-862a-004471ffe5dc.png)

This is one of the least expensive good quality AP2 setups that I've built. I've streamed for several hours on this build and have had zero issues and the Pi stays relatively cool. The toughest part of this project was soldering the gpio header pins. I'm hoping they will have the pre-soldered headers soon??

**Reply by mikebrady** [reply]
Thanks for the interesting writeup! (I made a few tiny layout changes — I hope that’s okay.)

**Reply by quantonian** [reply]
Thanks Mike! I'll make sure to use this format for future posts.
```
test
```

**Reply by noelhibbard** [reply]
Thanks for sharing. I have one of those Ikea ENEBY 20 BT speakers which has an AUX input that I want to AirPlay 2 enable and was planning to use this exact DAC. The ENEBY 20 has a battery compartment that I think will be just large enough to house a PiZero and this little DAC.

**Reply by Joeknx** [reply]
Thanks for the instructions! 
I used the same PCM5102A DAC from Aliexpress.
It works so far, the only problem is that the output is far to loud and the singnal is noisy.
Do you have any idea how to lower the hardware output? Thanks!

**Reply by Joeknx** [reply]
I found a solution for the loud signal:

add the following code to asound.conf

pcm.!default  {
 type plug     
 slave.pcm "speakerphat"
}

ctl.!default {
 type hw card 0
}

pcm.speakerphat {
  type softvol
  slave.pcm "plughw:0"
  control.name "Master"
  control.card 0
}

**Reply by Joeknx** [reply]
The noise in the signal sadly persists

**Reply by mikebrady** [reply]
@Joeknx, you can [set the maximum output level](https://github.com/mikebrady/shairport-sync/blob/master/ADVANCED%20TOPICS/GetTheBest.md#maximum-output-level).

Also, try outputting [directly to the output device](https://github.com/mikebrady/shairport-sync/blob/development/ADVANCED%20TOPICS/InitialConfiguration.md#specifying-the-output-device-and-mixer-control), bypassing any ALSA plugins.

Good luck!

**Reply by noelhibbard** [reply]
You can get the audio off the HDMI port. There are little HDMI dongles with headphone jacks on them on Amazon for $9. I've had better luck with those.

**Reply by Joeknx** [reply]
@mikebrady @noelhibbard 
thank you for the tips.
The noise sounds like a groud loop. Maybee I try the HDMI Adapter.

**Reply by Joeknx** [reply]
It had been the original! Raspberry power adapter which caused the "ground loop" sound.
With an iPhone power adapter its gone!

**Comment by strugatzki** [comment]
As this thread is about the Raspberry Pi Zero (not the Zero 2) and Zero 2's are not easy to get on the market currently, I'd like to confirm the observations above in a less technical language: _The Pi Zero indeed works well enough with AP2, with some limitations_ (which, if I understand Mike's summary explanation correctly, are not bugs, but result from the Pi Zero not having enough processing power for some tasks needed _sometimes_ for certain AP2 streams, and there's no way round it):

**1. Works well from iOS;** not sure if on iOS there is really a difference between "system audio" and selecting Airplay from within, say, the Music app or Spotify, but since I don't know what they do and Spotify certainly does something else than the control center when selecting AP from within the Spotify app, I always select Airplay from the control center, and like that it has been working perfectly for me. Indeed some events might make it unstable, such as reconnecting, but it plays in sync with no distortions and overall it's very good. From iOS, you also can select multiple Airplay targets, one of them being the Pi Zero (from a non-technical user perspective, this is all what AP2 is about), and I have no issues serving a heterogenous network consisting of two shairplay-sync clients (Pi 4 and Pi Zero), a Sonos AP2 mono speaker, and a HomePod Mini at the same time from an iPhone (I don't think the iOS version matters). The Zero can do some minor tasks alongside shairplay-sync, but your installation will have to be "almost dedicated". (Note that Hifiberry OS includes the AP2 development branch for shairport-sync, but will distort audio between tracks because of other things; you must do your own clean and bare bones install of Raspberry Pi OS Lite, as I believe Mike explains in the wonderful documentation to his superb piece of open source software).

**2. On MacOS and with standard tools, the Zero indeed will only play via system audio, never from Apple Music,** and again certain events might make it unstable, but usually you can solve these by reconnecting (note that it is not really system audio because macOS will still output system chimes to the local system), and it's fine. However, and that is the crucial thing, macOS system audio only can go to a single AP target, you can't do multi-target (it's been like that for ages, even for AP1 and iTunes -- multi-AP target from iTunes has been possible since a long, long time, even in AP1 times, if I recall correctly). You could only do that from within the Apple Music app, but that's where the Zero does not work .

**3. For macOS, there is a solution for multi-target that works:** Airfoil lets you push macOS system or even application audio to several AP targets at once, and that includes the Zero. It's not as convenient in terms of volume control and play/pause _(I hate lag in control),_ but it generally works good enough. (Not an ad for Airfoil, I knew it from ages ago and did not bother to research if there are alternatives). It would be interesting to find out if Airfoil sends an AP1 or AP2 stream (and if that, what kind)  to the Pi Zero, but tbh I don't care enough to investigate.

4. Summed up, the Zero works good enough, but I'll be getting myself a Zero 2 once prices are back to normal.

**Reply by charlesomer** [reply]
Hi @strugatzki, it's interesting you say it doesn't work with the Apple Music app from macOS. I've just tried this and it seems to connect okay? Could you connect but not play music?
